python310Packages.urllib3: remove pyopenssl as it is no longer recommended

dbarrosop opened the pull request:
###### Description of changes

Fixes #178954

Commit 292269900315d5b952a99ec95aa4ccfc099952e0  introduced a bunch of [optional dependencies](https://github.com/urllib3/urllib3/blob/main/pyproject.toml#L40-L56) as mandatory. This had the side effect of breaking urllib3 on darwin-aarch64 as pyopenssl isn't supported on that platform (https://github.com/NixOS/nixpkgs/issues/175875)

This PR introduces a couple of arguments to allow toggling optional dependencies on and off defaulting them to `true` with the exception of `darwin-aarch64` which disables by default the broken optional dependencies.

Note: I am unsure if this is the right branch, I just tried to replicate what was done on the PR that introduced the change I linked above.

Note2: I think ideally we'd have as root packages:

- python310Packages.urllib3 (no optional deps, default installation, same as pip)
- python310Packages.urllib3-full (all optional deps enabled)

but didn't want to introduce "big" changes, just fix urllib3 on `darwin-aarch64`

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

azuwis commented:
Pyopenssl is marked broken on aarch64-darwin (https://github.com/NixOS/nixpkgs/issues/175875),
but it's no longer recommended for urllib3 (https://github.com/urllib3/urllib3/blob/main/src/urllib3/contrib/pyopenssl.py).
Removing pyopenssl from dependencies on aarch64-darwin will save near 9000 packages.
For reference, Debian make pyopenssl `suggest` for urllib3 (https://packages.debian.org/sid/python3-urllib3),
which means pyopenssl will NOT be installed by default when installing urllib3.

SuperSandro2000 commented:
Wrong target branch. This should target master or staging.

SuperSandro2000 commented on pkgs/development/python-modules/urllib3/default.nix:
```diff
@@ -16,6 +17,8 @@
 , pytestCheckHook
 , tornado
 , trustme
+, enableBrotli ? true
+, enableSecure ? !(stdenv.isAarch64 && stdenv.isDarwin)
```
```suggestion
``` 
This is not allowed in pythonPackages and will cause hard to debug errors.

dbarrosop replied:
I removed the conditionals but happy to add them back defaulting them to `false` if we want to make them opt-in as suggested [here](https://github.com/NixOS/nixpkgs/pull/179159#discussion_r906826976)

SuperSandro2000 commented on pkgs/development/python-modules/urllib3/default.nix:
```diff
@@ -29,8 +32,9 @@ buildPythonPackage rec {
   };
 
   # FIXME: remove backwards compatbility hack
```
```suggestion
```
We should rather make both optional-dependencies opt-in. Since this will probably cause a massive fallout on linux this needs to go to staging or much preferable python-updates. But we currently don't have any python-updates run running, so this presumable needs to wait a little bit.

SuperSandro2000 commented:
> Removing pyopenssl from dependencies on aarch64-darwin will save near 9000 packages.

We shouldn't make such dependency packages platform depended if upstream has no explicit support for it. It will only cause extra maintenance for all other packages.

azuwis commented:
> > Removing pyopenssl from dependencies on aarch64-darwin will save near 9000 packages.
> 
> We shouldn't make such dependency packages platform depended if upstream has no explicit support for it. It will only cause extra maintenance for all other packages.

How about removing pyopenssl from urllib3 for all platforms?

From https://github.com/urllib3/urllib3/blob/main/src/urllib3/contrib/pyopenssl.py:

> Module for using pyOpenSSL as a TLS backend. This module was relevant before
the standard library ``ssl`` module supported SNI, but now that we've dropped
support for Python 2.7 all relevant Python versions support SNI so
**this module is no longer recommended**.

The pyopenssl urllib3 backend was mainly written for python 2.7, python3-urllib3 does NOT need pyopenssl.

SuperSandro2000 commented:
> How about removing pyopenssl from urllib3 for all platforms?

Yes, we should do that but it will cause a mass rebuild and I can't assess how big the breakage will be.

dbarrosop commented:
Thanks for the review, I updated my PR to:

1. target `staging`
2. simply remove `pyopenssl`
3. changed the commit message/PR title

After reading the comments I wasn't sure if we still wanted to make the optional dependencies opt-in, happy to make the changes if you also want to do that.

zmre commented:
(Just a note to say a ton of things have broken on my mac that I expect this to fix, so watching anxiously for this to make it through.)

SuperSandro2000 commented on pkgs/development/python-modules/urllib3/default.nix:
```diff
@@ -65,7 +64,7 @@ buildPythonPackage rec {
   passthru.optional-dependencies = {
     brotli = if isPyPy then [ brotlicffi ] else [ brotli ];
-    secure = [ certifi cryptography idna pyopenssl ];
+    secure = [ certifi cryptography idna ];
```
Do we have any we can link in urllib3 that mentions that pyopenssl is not required? If so we should link that here that pyopenssl is not added back later.

azuwis replied:
How about https://urllib3.readthedocs.io/en/latest/reference/contrib/pyopenssl.html

SuperSandro2000 commented:
> (Just a note to say a ton of things have broken on my mac that I expect this to fix, so watching anxiously for this to make it through.)

This will just fix packages that require urllib3 which are not that many. pyopenssl has still many other usages and this PR will probably not fix most of your issues.

azuwis commented:
I think we should also remove `cryptography` and `idna`, as https://urllib3.readthedocs.io/en/stable/reference/contrib/pyopenssl.html say they are dependancies of `pyopenssl`.

I'm not sure about `certifi` though.

dbarrosop commented:
I think I addressed all of your comments. Decided to comment the `secure` dependencies entirely as they don't seem relevant anymore. As mentioned by @azuwis, `cryptography` and `idna` are dependencies of `pyopenssl` and looking at `certifi` it seems they are only used in tests for `pyopenssl` [1](https://github.com/urllib3/urllib3/search?q=%22from+cryptography%22)[2](https://github.com/urllib3/urllib3/search?q=%22import+cryptography%22)

dbarrosop committed:
python310Packages.urllib3: remove pyopenssl as it is no longer recommended

fabaff commented on pkgs/development/python-modules/selenium/default.nix:
```diff
@@ -35,8 +35,7 @@ buildPythonPackage rec {
     trio
     trio-websocket
     urllib3
-  ] ++ urllib3.optional-dependencies.secure
```
`selenium` wants [`secure` and `socks`](https://github.com/SeleniumHQ/selenium/blob/trunk/py/setup.py#L67). Would help in the future if it's commented why `urllib3.optional-dependencies.secure` should not be added.

azuwis replied:
I think it's a wrong requirement, selenium does not use pyopenssl (https://github.com/SeleniumHQ/selenium/search?q=pyopenssl)

According to https://urllib3.readthedocs.io/en/latest/reference/contrib/pyopenssl.html, if pyopenssl is used, there should be some code like:

```
try:
    import urllib3.contrib.pyopenssl
    urllib3.contrib.pyopenssl.inject_into_urllib3()
except ImportError:
    pass
```

SuperSandro2000 replied:
It sill has a requirement in the requirements.txt and when you package things you don't double check everything https://github.com/SeleniumHQ/selenium/blob/trunk/py/requirements.txt#L20

jraygauthier reviewed (approved):
LGTM code-wise.

ylh commented:
for those of you at home on aarch64-darwin eagerly awaiting this getting merged, the following overlay should do the trick for now:

```
_: { python3, lib, ... }: {
  python3Packages = (python3.override {
    packageOverrides = _: { urllib3, ... }: {
      urllib3 = urllib3.overrideAttrs ({ passthru, ... }: {
        passthru = passthru // {
          optional-dependencies =
            lib.filterAttrs (n: _v: n != "secure")
            passthru.optional-dependencies;
        };
        propagatedBuildInputs = passthru.optional-dependencies.brotli;
      });
    };
  }).pkgs;
}
```

> This will just fix packages that require urllib3 which are not that many. pyopenssl has still many other usages and this PR will probably not fix most of your issues.

well, there's yt-dlp for starters

then, a cursory review of transitive build dependencies on urllib3 via sphinx on my machine alone:

- mpd
- qemu
- ghc
  - by extension, **literally every single haskell package**

i offer as friendly advice that dismissive handwaving in the comments of a PR that addresses breakage *you introduced* is a *monumentally* bad look, especially for a member of @NixOS

sternenseemann commented:
Workaround for GHC has been merged https://github.com/NixOS/nixpkgs/pull/179181

SuperSandro2000 commented:
> breakage _you introduced_

This just reflects upstreams support status and the issue that is open for 2.5 years. This already used many maintainer hours in some downstream packages where tests mysteriously broke because of this. Still I just used yet another hour for a proprietary OS that I don't even use. 

---

Since I can't push to your branch I need to open a new PR. Also reverting #179181 in its entirety there. So moving this to #179505

SuperSandro2000 commented:
The rebuild amount is only so low because of https://github.com/NixOS/nixpkgs/pull/178700 which means we couldn't have merged this into master.

dbarrosop commented:
Thanks for merging and for the thankless job that it is being an OSS maintainer, specially a project as massive and complex as nixpkgs.

SuperSandro2000 commented:
> Quick question, do you know when this will hit unstable approximately? I know it has to go through a few steps but I am not familiar with typical timeframes.

Since staging-next just merged (https://github.com/NixOS/nixpkgs/pull/178690) and I don't think there is a new PR yet, we could be lucky and based on the duration of the last staging-next run it could already be in 7 to 10 days but if problems happen it could also be 14 days.

zmre commented:
+1 on the thank you for the otherwise thankless job here. I was one of the ones anxious to see this hit and I know working on other people's platforms is a pain.  Thanks for getting it done.

Atemu commented:
This PR is broken:

```
error: attribute 'secure' missing

       at /Users/Atemu/Projects/nixpkgs/pkgs/development/python-modules/urllib3/default.nix:29:8:

           28|   propagatedBuildInputs = passthru.optional-dependencies.brotli
           29|     ++ passthru.optional-dependencies.secure;
             |        ^
           30|
(use '--show-trace' to show detailed location information)
```

That obviously needs to be removed aswell.

@SuperSandro2000 is it intended that the `socks` are not part of propagatedBuildInputs anymore here?

Atemu commented:
Ah, sorry. Fixed in f8d1cada7dbb7358614316e00e552e1a4cc9b441

SuperSandro2000 commented:
> Ah, sorry. Fixed in [f8d1cad](https://github.com/NixOS/nixpkgs/commit/f8d1cada7dbb7358614316e00e552e1a4cc9b441)

that PR took all the mass rebuilds and it is reason why this still must go to staging.

SuperSandro2000 commented:
> Though I think I would've preferred if an aarch64-darwin-only optionals flag made it to master first and staging then had a change to remove it for all platforms.

Then we would at least already have one merge conflict someone would need to fix again. 

> for at least a few weeks which isn't great.

pyopenssl is marked broken for aarch64-darwin since 11. May which is almost 6 weeks ago. Likely it will be fixed in a few days with #179844 already. I think we have that time now.

Atemu commented:
I meant to implement the darwin handling into master and then make a PR against staging where the optional darwin handling is replaced with the unconditional one/removal.

Now it's too late to do this without a merge conflict of course; I just would prefer to do it this way next time to reduce breakage time.

Workaround to master first, rebuild-heavy change that does it properly to staging.

> pyopenssl is marked broken for aarch64-darwin since 11. May which is almost 6 weeks ago.

Only in staging. It's only been broken on master (where everyone would notice) since the last staging-next.

SuperSandro2000 commented:
The PR that marked pyopenssl as broken went straight to master https://github.com/NixOS/nixpkgs/pull/172397 .

If the recent staging-next merge broke this due to dependency changes then this reflects aarch64-darwin in the support tier (https://github.com/NixOS/rfcs/blob/master/rfcs/0046-platform-support-tiers.md#tier-7-1) where it is the lowest possible.

Also with support tier 7 I don't think that extra work is justified. Arguable aarch64-darwin should probably be tier 4 or 3 rather than 7 but I still wouldn't like to go that extra round for that support tier.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-know-which-package-in-my-hm-config-depends-on-another/20163/9
